File: лаб4 дерево.java
java
import java.util.*;

class TreeNode {
    String value;
    List<TreeNode> children;
    
    public TreeNode(String value) {
        this.value = value;
        this.children = new ArrayList<>();
    }
    
    public void addChild(TreeNode child) {
        children.add(child);
    }
}

class Tree {
    TreeNode root;
    
    public List<String> findPath(String target) {
        List<String> path = new ArrayList<>();
        if (dfsFindPath(root, target, path)) {
            return path;
        }
        return new ArrayList<>();
    }
    
    private boolean dfsFindPath(TreeNode node, String target, List<String> path) {
        if (node == null) return false;
        
        path.add(node.value);
        if (node.value.equals(target)) return true;
        
        for (TreeNode child : node.children) {
            if (dfsFindPath(child, target, path)) return true;
        }
        
        path.remove(path.size() - 1);
        return false;
    }
}

File: лаб4 дерево.py
python
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.children = []
    
    def add_child(self, child_node):
        self.children.append(child_node)

class Tree:
    def __init__(self):
        self.root = None
    
    def find_path(self, target_value):
        """Поиск пути от корня до узла с заданным значением"""
        def dfs(node, path, target):
            if node is None:
                return None
            path.append(node.value)
            if node.value == target:
                return path.copy()
            for child in node.children:
                result = dfs(child, path, target)
                if result:
                    return result
            path.pop()
            return None
        
        return dfs(self.root, [], target_value) if self.root else None
